### ISSUE TITLE
Collection not typed when using EntityRepository::matching

### ISSUE THREAD

**Issue by ygottschalk** [weirdan/doctrine-psalm-plugin#63]
**TL;DR**: `EntityRepository<T>::matching` has return type `Collection`, should be `Collection<int, T>`

How to reproduce
--------------------
Code: https://psalm.dev/r/678bf40ce6
Error: `The type 'Doctrine\Common\Collections\Collection' is more general than the declared return type 'Doctrine\Common\Collections\Collection<int, Entity>' for MyRepository::foo`
Versions of my packages:
```
doctrine/collections                  1.6.6
doctrine/orm                          v2.7.3  
vimeo/psalm                           3.12.2
weirdan/doctrine-psalm-plugin         0.11.3
```
**Is it reproducible or is it just me?**

The case for adding `matching` to the stub
-----------------------------------------------
I started digging a bit and found that:
- the method declaring interface [\Doctrine\Common\Collections\Selectable](https://github.com/doctrine/collections/blob/master/lib/Doctrine/Common/Collections/Selectable.php) is 'perfectly typed'
- and has a 'perfectly typed' [stub ](https://github.com/weirdan/doctrine-psalm-plugin/blob/master/stubs/Collections.phpstub#L165-L175) (Is this necessary, since the folks over at doctrine already typed it?)

**BUT**:
- [EntityRepository#matching()](https://github.com/doctrine/orm/blob/2.7/lib/Doctrine/ORM/EntityRepository.php#L273-L286) seems to override the 'perfect type' from the interface (current stable v2.7.x)
- and it seems to not change in the future [see v2.8.x](https://github.com/doctrine/orm/blob/2.8.x/lib/Doctrine/ORM/EntityRepository.php#L284-L297), [see master](https://github.com/doctrine/orm/blob/master/lib/Doctrine/ORM/EntityRepository.php#L229-L240)

Therefore I'd like to add `matching` to the `EntityRepository`-stub. **I'm happy to provide a PR, just let me know**.

Thank you for taking the time and have a nice day!

**Comment by weirdan** [comment]
It's not just you. It can be reproduced with the following scenario (after adding necessary `use`s to the preamble):
```gherkin

  @EntityRepository::matching
  Scenario: matching is typed
    Given I have the following code
      """
      /** @return Collection<int, I> */
      function f(): Collection {
        return repo(I::class)->matching(Criteria::create());
      }
      """
    When I run Psalm
    Then I see no errors
```

> and has a 'perfectly typed' stub (Is this necessary, since the folks over at doctrine already typed it?)

The plugin supports fairly wide range of doctrine versions, including those that did not have Psalm types.

>  I'm happy to provide a PR, just let me know.

PRs are more than welcome!

> and it seems to not change in the future see v2.8.x, see master

You can also submit a PR against doctrine repo. It can even be backported to older branches (but of course it will only be available with new releases).
